### ISSUE TITLE
front end

### ISSUE THREAD

**Issue by rohithsenthil527-hash** [rohithsenthil527-hash/Road-accident-data-analysis-#1]
# ============================================================
# ROAD ACCIDENT DATA ANALYSIS — FULL PIPELINE
# ============================================================

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
import warnings
warnings.filterwarnings("ignore")

sns.set_style("whitegrid")
plt.rcParams["figure.dpi"] = 100

# ------------------------------------------------------------
# 1. LOAD DATA
# ------------------------------------------------------------
df = pd.read_csv("road_accidents.csv", parse_dates=["Date"])
print("Shape:", df.shape)
print(df.head())
print(df.info())

# ------------------------------------------------------------
# 2. CLEAN DATA
# ------------------------------------------------------------
print(df.isna().sum())

for col in ["Weather_Conditions", "Road_Surface_Conditions", "Junction_Detail"]:
    df[col] = df[col].fillna("Unknown")

df["Hour"] = pd.to_datetime(df["Time"], format="%H:%M").dt.hour
df["Month"] = df["Date"].dt.month_name()
df["Year"] = df["Date"].dt.year

# ------------------------------------------------------------
# 3. EXPLORATORY DATA ANALYSIS
# ------------------------------------------------------------
order = ["Slight", "Serious", "Fatal"]
colors = ["#4C9F70", "#E8A33D", "#C0392B"]

# 3.1 Severity distribution
ax = df["Accident_Severity"].value_counts().reindex(order).plot(
    kind="bar", color=colors, figsize=(7, 4))
ax.set_title("Accidents by Severity")
plt.tight_layout(); plt.savefig("01_severity_distribution.png"); plt.show()

# 3.2 Accidents by hour of day
fig, ax = plt.subplots(figsize=(9, 4))
df.groupby("Hour").size().plot(kind="line", marker="o", ax=ax, color="#2C3E50")
ax.set_title("Accidents by Hour of Day")
plt.tight_layout(); plt.savefig("02_accidents_by_hour.png"); plt.show()

# 3.3 Accidents by day of week
day_order = ["Monday","Tuesday","Wednesday","Thursday","Friday","Saturday","Sunday"]
fig, ax = plt.subplots(figsize=(8, 4))
df["Day_of_Week"].value_counts().reindex(day_order).plot(
    kind="bar", color="#3E7CB1", ax=ax)
ax.set_title("Accidents by Day of Week")
plt.tight_layout(); plt.savefig("03_accidents_by_day.png"); plt.show()

# 3.4 Weather conditions
fig, ax = plt.subplots(figsize=(8, 5))
order_w = df["Weather_Conditions"].value_counts().index
sns.countplot(data=df, y="Weather_Conditions", order=order_w, ax=ax, palette="Blues_r")
ax.set_title("Accidents by Weather Condition")
plt.tight_layout(); plt.savefig("04_accidents_by_weather.png"); plt.show()

# 3.5 Severity by urban/rural
cross = pd.crosstab(df["Urban_or_Rural_Area"], df["Accident_Severity"],
                     normalize="index").reindex(columns=order)
cross.plot(kind="bar", stacked=True, color=colors, figsize=(7, 4))
plt.title("Severity Proportion: Urban vs Rural")
plt.legend(title="Severity", bbox_to_anchor=(1.02, 1))
plt.tight_layout(); plt.savefig("05_urban_rural_severity.png"); plt.show()
print(cross.round(3))

# 3.6 Severity by speed limit
speed_sev = pd.crosstab(df["Speed_limit"], df["Accident_Severity"],
                         normalize="index").reindex(columns=order)
speed_sev.plot(kind="bar", stacked=True, color=colors, figsize=(8, 4))
plt.title("Severity Proportion by Speed Limit")
plt.legend(title="Severity", bbox_to_anchor=(1.02, 1))
plt.tight_layout(); plt.savefig("06_speed_severity.png"); plt.show()

# 3.7 Monthly trend
month_order = ["January","February","March","April","May","June","July",
                "August","September","October","November","December"]
fig, ax = plt.subplots(figsize=(9, 4))
df["Month"].value_counts().reindex(month_order).plot(
    kind="bar", color="#6C8EBF", ax=ax)
ax.set_title("Accidents by Month")
plt.tight_layout(); plt.savefig("07_monthly_trend.png"); plt.show()

# ------------------------------------------------------------
# 4. FEATURE ENGINEERING
# ------------------------------------------------------------
model_df = df.copy()
cat_cols = ["Road_Type", "Light_Conditions", "Weather_Conditions",
            "Road_Surface_Conditions", "Urban_or_Rural_Area",
            "Junction_Detail", "Vehicle_Type", "Day_of_Week"]

encoders = {}
for c in cat_cols:
    le = LabelEncoder()
    model_df[c] = le.fit_transform(model_df[c])
    encoders[c] = le

sev_le = LabelEncoder()
model_df["Severity_Label"] = sev_le.fit_transform(model_df["Accident_Severity"])

# ------------------------------------------------------------
# 5. TRAIN SEVERITY PREDICTION MODEL
# ------------------------------------------------------------
features = cat_cols + ["Speed_limit", "Number_of_Vehicles",
                        "Number_of_Casualties", "Hour"]
X = model_df[features]
y = model_df["Severity_Label"]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y)

clf = RandomForestClassifier(
    n_estimators=200, max_depth=10, random_state=42, class_weight="balanced")
clf.fit(X_train, y_train)
preds = clf.predict(X_test)

print("Test Accuracy:", round(accuracy_score(y_test, preds), 3))

# ------------------------------------------------------------
# 6. MODEL EVALUATION
# ------------------------------------------------------------
print(classification_report(y_test, preds, target_names=sev_le.classes_))

cm = confusion_matrix(y_test, preds)
fig, ax = plt.subplots(figsize=(5, 4))
sns.heatmap(cm, annot=True, fmt="d", cmap="Blues",
            xticklabels=sev_le.classes_, yticklabels=sev_le.classes_, ax=ax)
ax.set_xlabel("Predicted"); ax.set_ylabel("Actual")
ax.set_title("Confusion Matrix")
plt.tight_layout(); plt.savefig("08_confusion_matrix.png"); plt.show()

# Feature importance
importances = pd.Series(clf.feature_importances_, index=features).sort_values()
fig, ax = plt.subplots(figsize=(7, 5))
importances.plot(kind="barh", ax=ax, color="#2C3E50")
ax.set_title("Feature Importance (Severity Prediction)")
plt.tight_layout(); plt.savefig("09_feature_importance.png"); plt.show()
